keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+together with the Pod) and persistent (lifecycle managed separately from a Pod,
+with an allocated resource used for multiple different Pods).
+
+Several core Kubernetes components must be modified (see the
+[implementation](#implementation) section for details):
+- kube-apiserver (new API)
+- kube-controller-manager (new controller)
+- kube-scheduler (new builtin scheduler plugin)
+- kubelet (new third-party kubelet plugin type)
+
+Resources as described by this KEP are managed by third-party resource drivers that communicate with central
+Kubernetes components, in particular the kube-scheduler, by updating
+objects stored in the kube-apiserver. kube-scheduler only needs to be modified
+once to support dynamic resource allocation. Then multiple drivers from
+different vendors can be installed at the same time without making further
+changes to the scheduler.
+
+The scope of this KEP is smaller compared to two previous proposals (linked to
+below under non-goals) and therefore it is easier to implement:
+- kube-scheduler doesn't need to know about different kinds of resources
+  and therefore we don't need to define those in advance and later
+  update those definitions as new hardware appears.
+- There is no need to define standardized parameters for resources.
+
+Communication between the kubelet and the node part of the driver is
+handled through local Unix domain sockets and the kubelet plugin registration
+mechanism, using a new plugin type and a new gRPC interface.
+The container runtime uses the
+[Container Device Interface
+(CDI)](https://github.com/container-orchestrated-devices/container-device-interface)
+to expose the resources.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP addresses
+limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
+  reconfigure devices directly.
+
+  *Limitation*: Currently, it’s impossible to specify the desired
```

Comment:
> is the file intended to be sourced from the pod itself in this example? where is the source of truth for this file in this example?

The user provides the file or defines which one to use via a ResourceClaim parameter.

How that works in detail depends on the resource driver implementation. The file content could be given in a ConfigMap or referenced by a URL (probably more suitable if the data is large).

> is there an extension to the existing device plugin gRPC API that we could make to support this parameter 

We first would have to extend [ResourceList](https://github.com/kubernetes/api/blob/bfe324c0d64befbab416bd4063c0a38a6cda9334/core/v1/types.go#L5180) so that something can be specified there besides a numeric quantity. Is that really a direction we want to go in?